node/pkg/watchers/near/finalizer.go

```diff
@@ -0,0 +1,128 @@
+package near
+
+import (
+	"context"
+	"errors"
+
+	"github.com/certusone/wormhole/node/pkg/watchers/near/nearapi"
+	lru "github.com/hashicorp/golang-lru"
+	"go.uber.org/zap"
+)
+
+type Finalizer struct {
+	// internal cache of which blocks have been finalized, mapping blockHack => blockTimestamp.
+	// The timestamp is persisted because we'll need it again later.
+	// thread-safe
+	finalizedBlocksCache *lru.Cache
+	nearAPI              nearapi.NearApi
+	eventChan            chan eventType
+	mainnet              bool
+}
+
+func newFinalizer(eventChan chan eventType, nearAPI nearapi.NearApi, mainnet bool) Finalizer {
+	finalizedBlocksCache, _ := lru.New(workerCountTxProcessing * quequeSize)
+
+	return Finalizer{
+		finalizedBlocksCache,
+		nearAPI,
+		eventChan,
+		mainnet,
+	}
+}
+
+func (f Finalizer) isFinalizedCached(logger *zap.Logger, ctx context.Context, blockHash string) (bool, nearapi.BlockHeader) {
+	if err := nearapi.IsWellFormedHash(blockHash); err != nil {
+		// SECURITY defense-in-depth: check if block hash is well-formed
+		logger.Error("blockHash invalid", zap.String("error_type", "invalid_hash"), zap.String("blockHash", blockHash), zap.Error(err))
+		return false, nearapi.BlockHeader{}
+	}
+
+	if b, ok := f.finalizedBlocksCache.Get(blockHash); ok {
+		blockHeader := b.(nearapi.BlockHeader)
+		// SECURITY In blocks < 74473147 message timestamps were computed differently and we don't want to re-observe these messages
+		if !f.mainnet || blockHeader.Height > 74473147 {
+			return true, blockHeader
+		}
+	}
+
+	return false, nearapi.BlockHeader{}
+}
+
+// isFinalized() checks if a block is finalized by looking at the local cache first. If therer is an error during execution it returns false.
+// If it is not found in the cache, we walk forward up to nearBlockchainMaxGaps blocks by height,
+// starting at the block's height+2 and check if their value of "last_final_block" matches
+// the block in question.
+// we start at height+2 because NEAR consensus takes at least two blocks to reach finality.
+func (f Finalizer) isFinalized(logger *zap.Logger, ctx context.Context, queriedBlockHash string) (bool, nearapi.BlockHeader) {
+
+	logger.Debug("checking block finalization", zap.String("method", "isFinalized"), zap.String("parameters", queriedBlockHash))
+
+	// check cache first
+	if ok, block := f.isFinalizedCached(logger, ctx, queriedBlockHash); ok {
```

Comment:
Would it make sense to flip these two return values? When I first looked at it, I thought it was an error because map lookups return the ok as the second value.